mtags/src/main/scala-2/scala/meta/internal/pc/completions/MatchCaseCompletions.scala

```diff
@@ -68,78 +73,163 @@ trait MatchCaseCompletions { this: MetalsGlobal =>
-            context,
-            editRange,
-            autoImports
-          )
+
+        // Step 0: case for selector type
+        selectorSym.info match {
+          case NoType => ()
+          case _ =>
+            if (
+              !(selectorSym.isSealed &&
+                (selectorSym.isAbstract || selectorSym.isTrait))
+            )
+              visit(selectorSym, Identifier(selectorSym.name), Nil)
         }
-      }
 
-      // Step 1: walk through scope members.
-      metalsScopeMembers(pos).iterator
-        .foreach(m => visit(m.sym.dealiased, Identifier(m.sym.name), Nil))
+        // Step 1: walk through scope members.
+        metalsScopeMembers(pos).iterator
+          .foreach { m =>
+            val sym = m.sym.dealiased
+            val fsym = sym.dealiasedSingleType
+            val isValid = !parents.isParent(fsym) &&
+              (fsym.isCase ||
+                fsym.hasModuleFlag ||
+                fsym.isInstanceOf[TypeSymbol]) &&
+              parents.isSubClass(fsym, includeReverse = false)
+            if (isValid) visit(sym, Identifier(m.sym.name), Nil)
+          }
 
-      // Step 2: walk through known direct subclasses of sealed types.
-      val autoImport = autoImportPosition(pos, text)
-      parents.selector.typeSymbol.foreachKnownDirectSubClass { sym =>
-        autoImport match {
-          case Some(value) =>
-            val (shortName, edits) =
-              ShortenedNames.synthesize(sym, pos, context, value)
-            visit(sym, shortName, edits)
-          case scala.None =>
-            visit(sym, sym.fullNameSyntax, Nil)
+        // Step 2: walk through known direct subclasses of sealed types.
+        // In `List(foo).map { cas@@} we want to provide also `case (exhaustive)` completion
+        // which works like exhaustive match, so we need to collect only members from this step
+        val autoImport = autoImportPosition(pos, text)
+        val sealedDescs = mutable.Set.empty[Symbol]
+        selectorSym.foreachKnownDirectSubClass { sym =>
+          sealedDescs += sym.dealiased
+          autoImport match {
+            case Some(value) =>
+              val (shortName, edits) =
+                ShortenedNames.synthesize(sym, pos, context, value)
+              visit(sym, shortName, edits)
+            case None =>
+              visit(sym, sym.fullNameSyntax, Nil)
+          }
         }
+        val members = result.result()
+        val edits = members.map(_._2)
+        // In `List(foo).map { cas@@} we want to provide also `case (exhaustive)` completion
+        // which works like exhaustive match, so we need to collect only members from this step
+        if (includeExhaustive) {
+          def isSealedDesc(sym: Symbol) = {
+            sym.info match {
+              case tr: TypeRef => sealedDescs(tr.underlying.typeSymbol)
+              case _ => sealedDescs(sym)
+            }
+          }
+          val sortedMembers = sortSubclasses(members, selectorSym.tpe, source)
```

Comment:
In exhaustive case completion, we sort completion members the same way as for exhaustive match.
```scala
List(foo).map { case@@} 
List(foo).map { f => f match@@ } // ordering of cases is the same in both completions
```